### PR TITLE
add DataSourceStrategy with like support

### PR DESCRIPTION
**Pull request by adrian-wang**
## What changes were proposed in this pull request?

This is to enable spark push down STRING_LIKE operator

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by adrian-wang** [comment]
If we extend it, it would be much simplified, but need to merge with OAPStrategies. @yueawang

**Comment by chenghao-intel** [comment]
Maybe you can talk to @yueawang offline, and figure out if there any simple solution to push down more predictions into the data source, this is probably a common issues for the further optimization in OAP.

**Comment by adrian-wang** [comment]
Actually I only changed the `def translateFilters` in the big file. Should be easy if extend.

**Comment by yueawang** [comment]
Yes, If we create a base scan strategy class for OAP e.g. OapFileStrategy, we can control all these things in it.
This Daoyuan's change adds a case branch to spark which does not give a interface to register such case, that's why he need to extract out & modify the DataSourceStrategy from spark.
We can discuss later if we are able to implement the OapFileStrategy now.

File: src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+  private[this] def toCatalystRDD(relation: LogicalRelation, rdd: RDD[Row]): RDD[InternalRow] = {
+    toCatalystRDD(relation, relation.output, rdd)
+  }
+
+  /**
+   * Tries to translate a Catalyst [[Expression]] into data source [[Filter]].
+   *
+   * @return a `Some[Filter]` if the input [[Expression]] is convertible, otherwise a `None`.
+   */
+  protected[sql] def translateFilter(predicate: Expression): Option[Filter] = {
```
Comment (adrian-wang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: The function I want to modify is in an scala object, it seems I cannot derive an object from another...

**Comment by adrian-wang** [comment]
I'll merge this to unblock future work.

**Pull request merged**